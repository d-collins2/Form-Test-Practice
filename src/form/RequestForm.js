import React, { Component } from 'react';
import "./Form.css"
import { Row, Form } from 'react-bootstrap'
import TextInput from './TextInput.js'
import Select from './Select.js'
import TextArea from './TextArea.js'
import CheckBox from './CheckBox.js'
import Button from './Button.js'

class RequestForm extends Component{
    state = {
        formErrors: { firstName: '', lastName: '', email: '', serviceType: '', description: '' },
        firstName: '',
        firstNameValid: false,
        lastName: '',
        lastNameValid: false,
        email: '',
        emailValid: false,
        serviceType: '',
        serviceTypeValid: false,
        description: '',
        descriptionValid: false,
        subscribed: false,
        formValid: false
    }

    //Checks if the input field on the form is valid.
    validateField = (fieldName, value) => {
        let { firstNameValid, lastNameValid, emailValid, serviceTypeValid, descriptionValid, formErrors } = this.state
        switch(fieldName) {
            case 'firstName':
                firstNameValid = value.length >= 1;
                formErrors.firstName = firstNameValid ? '': 'First Name is empty.';
                break;
            case 'lastName':
                lastNameValid = value.length >= 1;
                formErrors.lastName = lastNameValid ? '': 'Last Name is empty.';
                break;
            case 'email':
                emailValid = value.match(/^([\w.%+-]+)@([\w-]+\.)+([\w]{2,})$/i);
                formErrors.email = emailValid ? '' : 'Email is invalid.';
                break;
            case 'serviceType':
                serviceTypeValid = value.length >= 1;
                formErrors.serviceType = serviceTypeValid ? '' : 'Service Type is invalid.';
                break;
            case 'description':
                descriptionValid = value.length >= 1;
                formErrors.description = descriptionValid ? '' : 'Description is empty.';
                break;
            default:
                break;
        }
        this.setState({
                        formErrors: formErrors,
                        firstNameValid: firstNameValid,
                        lastNameValid: lastNameValid,
                        emailValid: emailValid,
                        serviceTypeValid: serviceTypeValid,
                        descriptionValid: descriptionValid
                      }, this.validateForm);
    }

    // Checks if all fields are valid and allows the button to submit a request.
    validateForm() {
        const {
                firstNameValid,
                lastNameValid,
                emailValid,
                serviceTypeValid,
                descriptionValid,
                subscribed
              } = this.state
        this.setState({formValid:
                        firstNameValid &&
                        lastNameValid &&
                        emailValid &&
                        serviceTypeValid &&
                        descriptionValid &&
                        subscribed
                      });
    }

    // Sends a POST request which returns the response in a console.log.
    onSubmit = (event) => {
        event.preventDefault()
        const { firstName, lastName, email, serviceType, description } = this.state
        fetch('http://localhost:49567/api/assistance-requests', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'application/json'
            },
            body: JSON.stringify({
                assistance_request: {
                    'contact': {
                        'first_name': firstName,
                        'last_name': lastName,
                        'email': email
                    },
                    'service_type': serviceType,
                    'description': description
                }
            })
        })
        .then(res => res.json())
        .then(response => {
            console.log(response)
            this.setState({
                            firstName: '', lastName: '', email: '', serviceType: '', description: '' },
                            alert(response.message))
        })
    }

    // Sets state of the target input field and checks if its a valid input.
    handleChange = (event) => {
    let { name, type, checked, value } = event.target
    const newValue = type === 'checkbox' ? checked : value;
        this.setState({
            [name]: newValue },
            () => { this.validateField(name, newValue) }
        )
    }

    render() {
        const { firstName, lastName, email, serviceType, description} = this.state.formErrors
        console.log(this.state)
        return (
            <Row>
                <Form onSubmit={this.onSubmit} className="col-md-3 col-md-offset-4 form">
                    <h2 className="title">New Assistance Request</h2>
                    <TextInput
                        controlId={"FirstName"}
                        formError={firstName}
                        errorClass={this.errorClass}
                        handleChange={this.handleChange}
                        name={"firstName"}
                        placeholder={"First Name"}
                        value={this.state.firstName}/>
                    <TextInput
                        controlId={"LastName"}
                        errorClass={this.errorClass}
                        formError={lastName}
                        handleChange={this.handleChange}
                        name={"lastName"}
                        placeholder={"Last Name"}
                        value={this.state.lastName}/>
                    <TextInput
                        controlId={"Email"}
                        errorClass={this.errorClass}
                        formError={email}
                        handleChange={this.handleChange}
                        name={"email"}
                        placeholder={"Email Address"}
                        value={this.state.email}/>
                    <Select
                        controlId={"SelectType"}
                        errorClass={this.errorClass}
                        formError={serviceType}
                        handleChange={this.handleChange}
                        name={"serviceType"}
                        serviceTypes={this.props.serviceTypes}
                        value={this.state.serviceType}/>
                    <TextArea
                        controlId={"Description"}
                        errorClass={this.errorClass}
                        formError={description}
                        handleChange={this.handleChange}
                        name={"description"}
                        serviceTypes={this.props.description}
                        value={this.state.description}/>
                    <CheckBox
                        checked={this.state.subscribed}
                        handleChange={this.handleChange}
                        label="I hereby accept the terms of service for THE NETWORK and the Privacy Policy"
                        labelFor="subscribed"/>
                    <Button text={"Get Assitance"}/>
                </Form>
            </Row>
        );
    }
}


export default RequestForm;
