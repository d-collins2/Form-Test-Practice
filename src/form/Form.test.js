import React from 'react';
import { shallow, mount } from 'enzyme';
import RequestForm from './RequestForm.js';

describe('RequestForm', () => {
  it('should render correctly in "debug" mode', () => {
    const component = shallow(<RequestForm debug />);
    expect(component).toMatchSnapshot();
  });

  it('should render correctly with given strings', () => {
    const strings = ['one', 'two'];
    const component = shallow(<RequestForm list={strings} />);
    expect(component).toMatchSnapshot();
  });

  let component = shallow(<RequestForm/>);
  it('should render correctly with no props', () => {
    expect(component).toMatchSnapshot();
  });

  it('should find a submit button', () => {
    expect(component.find('Button').text()).toEqual('Get Assitance')
  });

  it('should not submit a request with out full validation', () => {
    expect(component.find('Button').simulate('click'))
  });

  component = mount(<RequestForm />);
  it('Should capture First Name correctly onChange', function(){
    const input = component.find('input').at(0);
    input.instance().value = 'Damon';
    input.simulate('change');
    expect(component.state().firstName).toEqual('Damon');
  })

  it('Should capture Last Name correctly onChange', function(){
    const input = component.find('input').at(1);
    input.instance().value = 'Collins';
    input.simulate('change');
    expect(component.state().lastName).toEqual('Collins');
  })

  it('Should capture Email Address correctly onChange', function(){
    const component = mount(<RequestForm />);
    const input = component.find('input').at(2);
    input.instance().value = 'test@gmail.com';
    input.simulate('change');
    expect(component.state().email).toEqual('test@gmail.com');
  })

  it('Should capture Service Type correctly onChange', function(){
    const input = component.find('select').at(0);
    const optionEnglish = component.find('option').at(0);
    optionEnglish.instance().value = "Benefits";
    input.simulate('change')
    expect(component.find('select').at(0).props().value).toEqual('Benefits');
  })

  it('should submit a request if all validations are true', () => {
    const state = {
      firstName:'hello',
      lastName:'world',
      email:"test@test.com",
      serviceType:"Benefits",
      description:"Need alot of help",
    }

    const expectedArg = "lirstName: hello, lastName: world, email: test@test.com, serviceType: Benefits, description: Need alot of help";
    window.alert = jest.fn();
    component.setState(state)
    component.find('form').simulate('submit')
    expect(window.alert).toHaveBeenCalledWith(expectedArg);
  });

});
