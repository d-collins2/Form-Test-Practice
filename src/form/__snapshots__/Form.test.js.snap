// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestForm should render correctly in "debug" mode 1`] = `ShallowWrapper {}`;

exports[`RequestForm should render correctly with given strings 1`] = `ShallowWrapper {}`;

exports[`RequestForm should render correctly with no props 1`] = `ReactWrapper {}`;
